corehq/apps/app_manager/management/commands/convert_mobile_ucr_refs.py

```diff
@@ -0,0 +1,78 @@
+import json
+import re
+import uuid
+from copy import deepcopy
+
+from django.core.management.base import BaseCommand
+
+from corehq.apps.app_manager.dbaccessors import get_app
+from corehq.apps.app_manager.exceptions import AppValidationError
+from corehq.apps.app_manager.tasks import prune_auto_generated_builds
+
+# Mobile UCR report references:
+# V1: `instance('reports')/reports/report[@id='xxxxxxx']`
+# V2: `instance('commcare-reports:xxxxxxx')`
+V1_REPORT_REFS = r"instance\('reports'\)/reports/report\[@id='([^']+)'\]"
+V2_REPORT_REFS = r"instance\('commcare-reports:\1'\)"
+
+# Mobile UCR column references:
+# V1: `column[@id = 'column_id']`
+# V2: `column_id`
+V1_COLUMN_REFS = r"column\[@id.?=.?'([^']+)'\]"
+V2_COLUMN_REFS = r'\1'
+
+
+class Command(BaseCommand):
+    help = 'Convert Mobile UCR references from V1 to V2'
+
+    def add_arguments(self, parser):
+        parser.add_argument('domain')
+        parser.add_argument('app_id')
+
+    def handle(self, domain, app_id, **options):
+        # Based on ApplicationBase.make_build()
+        v1_app = get_app(domain, app_id)
+        assert v1_app.copy_of is None
+
+        app_dict = _copy_app_dict(v1_app.to_json())
```

Comment:
This line deep-copies `v1_app`, dropping the keys we don't want. But in [`ApplicationBase.make_build()`](https://github.com/dimagi/commcare-hq/blob/791ada683322e0253e2190221a065d728873bec8/corehq/apps/app_manager/models.py#L4431-L4438) we only do this if the Couch view doesn't return any copies:
```python

        copies = cls.view(
            'app_manager/applications',
            key=[self.domain, self._id, self.version],
            include_docs=True,
            limit=1
        ).all()
        if copies:
            copy = copies[0]
        else:
            copy = deepcopy(self.to_json())
```

@orangejenny is `v1_app.to_json()` the app that we want? Or should we be fetching the app from the Couch view, and then doing something like ...
```python
        if copies:
            v1_app_copy = copies[0]
            app_dict = _copy_app_dict(v1_app_copy.to_json())
        else:
            app_dict = _copy_app_dict(v1_app.to_json())
```
?

Reply:
I don't know the answer to your direct question, although I'd be inclined to use the block of code you copied above rather than using `to_json` (when in doubt, copy the code that's used in production).

This isn't going to pull form XML, which I assume is where a lot of the references are. `Application.export_json` *does* pull in form XML. That's what we use to export apps for re-importing on another server. 

Doing an export, replace, and then re-import would be probably the cleanest/safest technical way to do this, since it's a well-tested process. But I assume that creating a new app isn't acceptable for the project. I wonder if it's possible to export, re-import as a new app, make a new build of the new app, and then update `copy_of` and `version` on that build to point to the pre-existing app. What do you think?

I thought we had done a script like this back when ICDS-CAS was migrated to UCR v2, but it [looks like](https://docs.google.com/document/d/13fkfPWN56lXk620g9sdM97WeuJWTdckjJJarkJfOo4g/edit?tab=t.0) what we did then was run an automated process to *find* all the references, and then the delivery team made the updates in HQ. I'm guessing a lot of them were in forms and for those ones they did a find and replace on each form's XML, which doesn't sound that burdensome. That doc also discusses mobile UCR v1.5, which helped with that process. I think that would be safer, although if this is being done for a single partner, I suppose it's up to them how much risk they're willing to take on.